{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries that are required to run your project\n",
    "# You are allowed to add more libraries as you need\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import spearmanr\n",
    "import os\n",
    "import tqdm\n",
    "import genomicranges as gr"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Information about data: \n",
    "- CAGE-train (\"data/CAGE-train\"): includes information about gene location and expression. X1 and X2 has train/val info/y, X3 only has test info that we should test on (only testing on chr1). \n",
    "- DNase-seq: information on chromatin accessibility (reads only in open chromatin)\n",
    "- bed, bigwig: genomic regions and associated annotations \n",
    "- do not use H3K9me3, distribution too different among X1, X2, X3 \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Work Package 1.1 - Modeling Choices & Data Pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: \n",
    "# bed and bigwig files contain signals of all chromosomes (including sex chromosomes).\n",
    "# Training and validation split based on chromosomes has been done for you. \n",
    "# However, you can resplit the data in any way you want.\n",
    "\n",
    "current_path = os.getcwd()\n",
    "path_data = current_path + \"/data/\"\n",
    "path_test = current_path + \"/data/CAGE-train/X3_test_info.tsv\"   \n",
    "test_genes = pd.read_csv(path_test, sep='\\t')\n",
    "test_genes_gr = test_genes.rename(columns={'chr': 'seqnames', 'TSS_start': 'starts', 'TSS_end': 'ends'})\n",
    "\n",
    "# load bed files from separate files in a loop and store them in a list\n",
    "bedfiles = ['DNase-bed', 'H3K4me1-bed', 'H3K4me3-bed', 'H3K9me3-bed', 'H3K27ac-bed', 'H3K27me3-bed', 'H3K36me3-bed']\n",
    "bed_files_x1 = {}\n",
    "bed_files_x2 = {}\n",
    "bed_files_x3 = {}\n",
    "\n",
    "for i in bedfiles:\n",
    "    bed_files_x1[str(i)] = pd.read_csv(path_data + i + '/X1.bed', sep='\\t', header=None)\n",
    "    bed_files_x2[str(i)] = pd.read_csv(path_data + i + '/X2.bed', sep='\\t', header=None)\n",
    "    bed_files_x3[str(i)] = pd.read_csv(path_data + i + '/X3.bed', sep='\\t', header=None)\n",
    "\n",
    "# loading gene expression \n",
    "gex_path = 'CAGE-train'\n",
    "x1_train_info = pd.read_csv(path_data + gex_path + '/X1_train_info.tsv', sep='\\t')\n",
    "x1_train_info_gr = x1_train_info.rename(columns={'chr': 'seqnames', 'TSS_start': 'starts', 'TSS_end': 'ends'})\n",
    "\n",
    "x1_train_y = pd.read_csv(path_data + gex_path + '/X1_train_y.tsv', sep='\\t')\n",
    "x1_val_y = pd.read_csv(path_data + gex_path + '/X1_val_y.tsv', sep='\\t')\n",
    "\n",
    "x2_train_info = pd.read_csv(path_data + gex_path + '/X2_train_info.tsv', sep='\\t')\n",
    "x2_train_info_gr = x2_train_info.rename(columns={'chr': 'seqnames', 'TSS_start': 'starts', 'TSS_end': 'ends'})\n",
    "\n",
    "x2_train_y = pd.read_csv(path_data + gex_path + '/X2_train_y.tsv', sep='\\t')\n",
    "x2_val_y = pd.read_csv(path_data + gex_path + '/X2_val_y.tsv', sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "cannot concatenate object of type '<class 'genomicranges.GenomicRanges.GenomicRanges'>'; only Series and DataFrame objs are valid",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/vq/08wnbx9x3sg3mkwxf3r_4p_h0000gn/T/ipykernel_28288/3301502611.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# row bind x1 and x2 and x3 info\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mx1_x2_x3_train_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mx1_train_info_gr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx2_train_info_gr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtest_genes_gr\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mx1_x2_x3_train_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGenomicRanges\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx1_x2_x3_train_info\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mtest_genes_gr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGenomicRanges\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtest_genes_gr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    309\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m                 )\n\u001b[0;32m--> 311\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    313\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/concat.py\u001b[0m in \u001b[0;36mconcat\u001b[0;34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[0m\n\u001b[1;32m    292\u001b[0m     \u001b[0mValueError\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIndexes\u001b[0m \u001b[0mhave\u001b[0m \u001b[0moverlapping\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'a'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    293\u001b[0m     \"\"\"\n\u001b[0;32m--> 294\u001b[0;31m     op = _Concatenator(\n\u001b[0m\u001b[1;32m    295\u001b[0m         \u001b[0mobjs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    296\u001b[0m         \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/reshape/concat.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, objs, axis, join, keys, levels, names, ignore_index, verify_integrity, copy, sort)\u001b[0m\n\u001b[1;32m    382\u001b[0m                     \u001b[0;34m\"only Series and DataFrame objs are valid\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    383\u001b[0m                 )\n\u001b[0;32m--> 384\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    385\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    386\u001b[0m             \u001b[0mndims\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: cannot concatenate object of type '<class 'genomicranges.GenomicRanges.GenomicRanges'>'; only Series and DataFrame objs are valid"
     ]
    }
   ],
   "source": [
    "# row bind x1 and x2 and x3 info \n",
    "x1_x2_x3_train_info = pd.concat([x1_train_info_gr, x2_train_info_gr, test_genes_gr])\n",
    "x1_x2_x3_train_info = gr.GenomicRanges(x1_x2_x3_train_info)\n",
    "\n",
    "test_genes_gr = gr.GenomicRanges(test_genes_gr)\n",
    "x1_train_info_gr = gr.GenomicRanges(x1_train_info_gr)\n",
    "x2_train_info_gr = gr.GenomicRanges(x2_train_info_gr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create promoter and enhancer regions in pandas dataframe \n",
    "rF = 5000\n",
    "promoter = x1_x2_x3_train_info.copy()\n",
    "promoter.starts = promoter.starts - rF\n",
    "promoter.ends = promoter.starts + rF\n",
    "promoter = gr.GenomicRanges(promoter)\n",
    "\n",
    "enhancer = x1_x2_x3_train_info.copy()\n",
    "enhancer.starts = enhancer.starts - 500000\n",
    "enhancer.ends = enhancer.starts + 500000\n",
    "enhancer = gr.GenomicRanges(enhancer)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>starts</th>\n",
       "      <th>ends</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>...</th>\n",
       "      <th>starts</th>\n",
       "      <th>ends</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>starts</th>\n",
       "      <th>ends</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SLC20A1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>112645939</td>\n",
       "      <td>112663825</td>\n",
       "      <td>112658362</td>\n",
       "      <td>112658412</td>\n",
       "      <td>+</td>\n",
       "      <td>SLC20A1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>112645939</td>\n",
       "      <td>...</td>\n",
       "      <td>112653362</td>\n",
       "      <td>112658362</td>\n",
       "      <td>+</td>\n",
       "      <td>SLC20A1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>112645939</td>\n",
       "      <td>112663825</td>\n",
       "      <td>112158362</td>\n",
       "      <td>112658362</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C11orf58</td>\n",
       "      <td>chr11</td>\n",
       "      <td>16613132</td>\n",
       "      <td>16758340</td>\n",
       "      <td>16738643</td>\n",
       "      <td>16738693</td>\n",
       "      <td>+</td>\n",
       "      <td>C11orf58</td>\n",
       "      <td>chr11</td>\n",
       "      <td>16613132</td>\n",
       "      <td>...</td>\n",
       "      <td>16733643</td>\n",
       "      <td>16738643</td>\n",
       "      <td>+</td>\n",
       "      <td>C11orf58</td>\n",
       "      <td>chr11</td>\n",
       "      <td>16613132</td>\n",
       "      <td>16758340</td>\n",
       "      <td>16238643</td>\n",
       "      <td>16738643</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28224886</td>\n",
       "      <td>28233487</td>\n",
       "      <td>28225263</td>\n",
       "      <td>28225313</td>\n",
       "      <td>+</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28224886</td>\n",
       "      <td>...</td>\n",
       "      <td>28220263</td>\n",
       "      <td>28225263</td>\n",
       "      <td>+</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6</td>\n",
       "      <td>28224886</td>\n",
       "      <td>28233487</td>\n",
       "      <td>27725263</td>\n",
       "      <td>28225263</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CD19</td>\n",
       "      <td>chr16</td>\n",
       "      <td>28931965</td>\n",
       "      <td>28939342</td>\n",
       "      <td>28931956</td>\n",
       "      <td>28932006</td>\n",
       "      <td>+</td>\n",
       "      <td>CD19</td>\n",
       "      <td>chr16</td>\n",
       "      <td>28931965</td>\n",
       "      <td>...</td>\n",
       "      <td>28926956</td>\n",
       "      <td>28931956</td>\n",
       "      <td>+</td>\n",
       "      <td>CD19</td>\n",
       "      <td>chr16</td>\n",
       "      <td>28931965</td>\n",
       "      <td>28939342</td>\n",
       "      <td>28431956</td>\n",
       "      <td>28931956</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TMEM123</td>\n",
       "      <td>chr11</td>\n",
       "      <td>102396332</td>\n",
       "      <td>102470384</td>\n",
       "      <td>102452789</td>\n",
       "      <td>102452839</td>\n",
       "      <td>-</td>\n",
       "      <td>TMEM123</td>\n",
       "      <td>chr11</td>\n",
       "      <td>102396332</td>\n",
       "      <td>...</td>\n",
       "      <td>102447789</td>\n",
       "      <td>102452789</td>\n",
       "      <td>-</td>\n",
       "      <td>TMEM123</td>\n",
       "      <td>chr11</td>\n",
       "      <td>102396332</td>\n",
       "      <td>102470384</td>\n",
       "      <td>101952789</td>\n",
       "      <td>102452789</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979</th>\n",
       "      <td>BRINP2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>177170958</td>\n",
       "      <td>177282422</td>\n",
       "      <td>177256780</td>\n",
       "      <td>177256830</td>\n",
       "      <td>+</td>\n",
       "      <td>BRINP2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>177170958</td>\n",
       "      <td>...</td>\n",
       "      <td>177251780</td>\n",
       "      <td>177256780</td>\n",
       "      <td>+</td>\n",
       "      <td>BRINP2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>177170958</td>\n",
       "      <td>177282422</td>\n",
       "      <td>176756780</td>\n",
       "      <td>177256780</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1980</th>\n",
       "      <td>PPIE</td>\n",
       "      <td>chr1</td>\n",
       "      <td>39692182</td>\n",
       "      <td>39763914</td>\n",
       "      <td>39738878</td>\n",
       "      <td>39738928</td>\n",
       "      <td>+</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>chr1</td>\n",
       "      <td>39692182</td>\n",
       "      <td>...</td>\n",
       "      <td>39733878</td>\n",
       "      <td>39738878</td>\n",
       "      <td>+</td>\n",
       "      <td>PPIE</td>\n",
       "      <td>chr1</td>\n",
       "      <td>39692182</td>\n",
       "      <td>39763914</td>\n",
       "      <td>39238878</td>\n",
       "      <td>39738878</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>CRB1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>197268204</td>\n",
       "      <td>197478455</td>\n",
       "      <td>197268277</td>\n",
       "      <td>197268327</td>\n",
       "      <td>+</td>\n",
       "      <td>CRB1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>197268204</td>\n",
       "      <td>...</td>\n",
       "      <td>197263277</td>\n",
       "      <td>197268277</td>\n",
       "      <td>+</td>\n",
       "      <td>CRB1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>197268204</td>\n",
       "      <td>197478455</td>\n",
       "      <td>196768277</td>\n",
       "      <td>197268277</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1982</th>\n",
       "      <td>TRIM33</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114392790</td>\n",
       "      <td>114511203</td>\n",
       "      <td>114406940</td>\n",
       "      <td>114406967</td>\n",
       "      <td>-</td>\n",
       "      <td>TRIM33</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114392790</td>\n",
       "      <td>...</td>\n",
       "      <td>114401940</td>\n",
       "      <td>114406940</td>\n",
       "      <td>-</td>\n",
       "      <td>TRIM33</td>\n",
       "      <td>chr1</td>\n",
       "      <td>114392790</td>\n",
       "      <td>114511203</td>\n",
       "      <td>113906940</td>\n",
       "      <td>114406940</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>SPATA1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>84506300</td>\n",
       "      <td>84566194</td>\n",
       "      <td>84506396</td>\n",
       "      <td>84506418</td>\n",
       "      <td>+</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>84506300</td>\n",
       "      <td>...</td>\n",
       "      <td>84501396</td>\n",
       "      <td>84506396</td>\n",
       "      <td>+</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>84506300</td>\n",
       "      <td>84566194</td>\n",
       "      <td>84006396</td>\n",
       "      <td>84506396</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30604 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name seqnames gene_start   gene_end     starts       ends  strand  \\\n",
       "0      SLC20A1     chr2  112645939  112663825  112658362  112658412       +   \n",
       "1     C11orf58    chr11   16613132   16758340   16738643   16738693       +   \n",
       "2       ZSCAN9     chr6   28224886   28233487   28225263   28225313       +   \n",
       "3         CD19    chr16   28931965   28939342   28931956   28932006       +   \n",
       "4      TMEM123    chr11  102396332  102470384  102452789  102452839       -   \n",
       "...        ...      ...        ...        ...        ...        ...     ...   \n",
       "1979    BRINP2     chr1  177170958  177282422  177256780  177256830       +   \n",
       "1980      PPIE     chr1   39692182   39763914   39738878   39738928       +   \n",
       "1981      CRB1     chr1  197268204  197478455  197268277  197268327       +   \n",
       "1982    TRIM33     chr1  114392790  114511203  114406940  114406967       -   \n",
       "1983    SPATA1     chr1   84506300   84566194   84506396   84506418       +   \n",
       "\n",
       "     gene_name seqnames gene_start  ...     starts       ends  strand  \\\n",
       "0      SLC20A1     chr2  112645939  ...  112653362  112658362       +   \n",
       "1     C11orf58    chr11   16613132  ...   16733643   16738643       +   \n",
       "2       ZSCAN9     chr6   28224886  ...   28220263   28225263       +   \n",
       "3         CD19    chr16   28931965  ...   28926956   28931956       +   \n",
       "4      TMEM123    chr11  102396332  ...  102447789  102452789       -   \n",
       "...        ...      ...        ...  ...        ...        ...     ...   \n",
       "1979    BRINP2     chr1  177170958  ...  177251780  177256780       +   \n",
       "1980      PPIE     chr1   39692182  ...   39733878   39738878       +   \n",
       "1981      CRB1     chr1  197268204  ...  197263277  197268277       +   \n",
       "1982    TRIM33     chr1  114392790  ...  114401940  114406940       -   \n",
       "1983    SPATA1     chr1   84506300  ...   84501396   84506396       +   \n",
       "\n",
       "     gene_name seqnames gene_start   gene_end     starts       ends  strand  \n",
       "0      SLC20A1     chr2  112645939  112663825  112158362  112658362       +  \n",
       "1     C11orf58    chr11   16613132   16758340   16238643   16738643       +  \n",
       "2       ZSCAN9     chr6   28224886   28233487   27725263   28225263       +  \n",
       "3         CD19    chr16   28931965   28939342   28431956   28931956       +  \n",
       "4      TMEM123    chr11  102396332  102470384  101952789  102452789       -  \n",
       "...        ...      ...        ...        ...        ...        ...     ...  \n",
       "1979    BRINP2     chr1  177170958  177282422  176756780  177256780       +  \n",
       "1980      PPIE     chr1   39692182   39763914   39238878   39738878       +  \n",
       "1981      CRB1     chr1  197268204  197478455  196768277  197268277       +  \n",
       "1982    TRIM33     chr1  114392790  114511203  113906940  114406940       -  \n",
       "1983    SPATA1     chr1   84506300   84566194   84006396   84506396       +  \n",
       "\n",
       "[30604 rows x 21 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "x1_x2_x3_train_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "reach = 5000\n",
    "\n",
    "# extracting features from bed files and gene info data\n",
    "def extract_features(input_bed_files, gene_info):\n",
    "    # iterate through rows in gene_info \n",
    "    # return pandas dataframe with features\n",
    "\n",
    "    # create empty dataframe\n",
    "    features = pd.DataFrame()\n",
    "    for index, row in gene_info.iterrows():\n",
    "        chr = row['chr']\n",
    "        gene_length = row['gene_end'] - row['gene_start']\n",
    "        tss_start = row['TSS_start'] - reach\n",
    "        tss_end = row['TSS_start'] + reach\n",
    "        # loop through bed files\n",
    "        for j in bedfiles:\n",
    "            print(j)\n",
    "            bed_files = input_bed_files[j]\n",
    "            # filter bed files by chromosome\n",
    "            bed_files_filt = bed_files.loc[bed_files[0] == chr]\n",
    "            # print(bed_files_filt)\n",
    "            # select for rows that are within reach of TSS\n",
    "            bed_files_final = bed_files_filt[(bed_files_filt[1] >= tss_start) & (bed_files_filt[2] <= tss_end)]\n",
    "            # if bed_files_final not empty\n",
    "            if bed_files_final.size > 0:\n",
    "                if j == 'DNase-bed':\n",
    "                    score = bed_files_final[6]\n",
    "                else:\n",
    "                    score = bed_files_final[4]\n",
    "                    print(score)\n",
    "                # peaks location \n",
    "                peak_start = bed_files_final[1]\n",
    "                peak_end = bed_files_final[2]\n",
    "\n",
    "                # sum of score\n",
    "                sum_score = score.sum()\n",
    "                # average score\n",
    "                avg_score = score.mean()\n",
    "                # max score\n",
    "                max_score = score.max()\n",
    "                # number of peaks\n",
    "                num_peaks = len(score)\n",
    "\n",
    "                # distance of tss midpoint to nearest peak midpoint\n",
    "                peak_midpoint = (peak_start + peak_end) / 2\n",
    "                tss_midpoint = (tss_start + tss_end) / 2\n",
    "                if num_peaks == 0:\n",
    "                    # arbtiary large distance if no peaks\n",
    "                    dist_to_peak = 2 * reach\n",
    "                else:\n",
    "                    # find index of peak which has midpoint closest to tss midpoint\n",
    "                    closest_peak = np.argmin(np.abs(peak_midpoint - tss_midpoint))\n",
    "                    # distance of tss midpoint to nearest peak midpoint\n",
    "                    dist_to_peak = np.abs(peak_midpoint[closest_peak] - tss_midpoint)\n",
    "            else:\n",
    "                sum_score = 0\n",
    "                avg_score = 0\n",
    "                max_score = 0\n",
    "                num_peaks = 0\n",
    "                dist_to_peak = 2 * reach\n",
    "        features = features.append({'gene_length': gene_length, 'sum_score': sum_score, 'avg_score': avg_score, 'max_score': max_score, \n",
    "                                        'num_peaks': num_peaks, 'dist_to_peak': dist_to_peak}, ignore_index=True)\n",
    "        print(features)\n",
    "    features.set_index(gene_info['gene_id'], inplace=True)\n",
    "    return features\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNase-bed\n",
      "H3K4me1-bed\n",
      "47457    46\n",
      "47458    50\n",
      "47459    48\n",
      "Name: 4, dtype: int64\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3360\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3361\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.Int64HashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.Int64HashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 2",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/vq/08wnbx9x3sg3mkwxf3r_4p_h0000gn/T/ipykernel_28288/3729101842.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# extract features for each dataset\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mx1_train_features\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mextract_features\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbed_files_x1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx1_train_info\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# print(bed_files_x1['H3K4me3-bed'])\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/vq/08wnbx9x3sg3mkwxf3r_4p_h0000gn/T/ipykernel_28288/2829995655.py\u001b[0m in \u001b[0;36mextract_features\u001b[0;34m(input_bed_files, gene_info)\u001b[0m\n\u001b[1;32m     52\u001b[0m                     \u001b[0mclosest_peak\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mabs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpeak_midpoint\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mtss_midpoint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     53\u001b[0m                     \u001b[0;31m# distance of tss midpoint to nearest peak midpoint\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 54\u001b[0;31m                     \u001b[0mdist_to_peak\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mabs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpeak_midpoint\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mclosest_peak\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mtss_midpoint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     55\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m                 \u001b[0msum_score\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    940\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    941\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mkey_is_scalar\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 942\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    943\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    944\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_hashable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m_get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m   1049\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1050\u001b[0m         \u001b[0;31m# Similar to Index.get_value, but we do not fall back to positional\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1051\u001b[0;31m         \u001b[0mloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1052\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_values_for_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1053\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3361\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3363\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3364\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3365\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhasnans\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 2"
     ]
    }
   ],
   "source": [
    "# extract features for each dataset\n",
    "x1_train_features = extract_features(bed_files_x1, x1_train_info)\n",
    "\n",
    "# print(bed_files_x1['H3K4me3-bed'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define function to subtract and add to tss region\n",
    "def tss_region(tss, start, end):\n",
    "    tss['TSS_start'] = tss['TSS_start'] - start\n",
    "    tss['TSS_end'] = tss['TSS_end'] + end\n",
    "    return tss\n",
    "\n",
    "# define a function to loop through bed files and extract signal sum in TSS region\n",
    "def extract_signal(bed_files, tss, features):\n",
    "    for i in tqdm.tqdm(bedfiles, desc= \"looping through bedfiles\"):\n",
    "        bed = bed_files[i]\n",
    "        chr = bed[0]\n",
    "        start = bed[1]\n",
    "        end = bed[2]\n",
    "        \n",
    "        if i == 'DNase-bed':\n",
    "            signal = bed[6]\n",
    "        else:\n",
    "            signal = bed[4]\n",
    "        for j in tss.index:\n",
    "            if any((chr == tss.loc[j, 'chr']) & (start >= tss.loc[j, 'TSS_start']) & (end <= tss.loc[j, 'TSS_end'])):\n",
    "                tss_signal = signal[(chr == tss.loc[j, 'chr']) & (start >= tss.loc[j, 'TSS_start']) & (end <= tss.loc[j, 'TSS_end'])].sum()\n",
    "            else:\n",
    "                tss_signal = 0\n",
    "            features.loc[j, i] = tss_signal\n",
    "        print(\"finished \" + i)\n",
    "        print(features.loc[:,i])\n",
    "    features = pd.DataFrame(features)\n",
    "    features.set_index(tss.index)\n",
    "    return features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define tss region\n",
    "tss_train = pd.DataFrame(x1_train_info[['gene_name', 'chr', 'TSS_start', 'TSS_end']])\n",
    "tss_train = tss_train.set_index('gene_name')\n",
    "tss_val = pd.DataFrame(x1_val_info[['gene_name', 'chr', 'TSS_start', 'TSS_end']])\n",
    "tss_val = tss_val.set_index('gene_name')\n",
    "tss_train = tss_region(tss_train, 5000, 5000)\n",
    "tss_val = tss_region(tss_val, 5000, 5000)\n",
    "\n",
    "tss_x3 = test_genes[['gene_name', 'chr', 'TSS_start', 'TSS_end']]\n",
    "tss_x3 = tss_x3.set_index('gene_name')\n",
    "tss_x3 = tss_region(tss_x3, 5000, 5000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "looping through bedfiles:   0%|          | 0/7 [00:13<?, ?it/s]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/huijeong/Library/Mobile Documents/com~apple~CloudDocs/ETH Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb Cell 7\u001b[0m in \u001b[0;36m<cell line: 8>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m features_x2_train \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame()\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m features_x2_val \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame()\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m features_x1_train \u001b[39m=\u001b[39m extract_signal(bed_files_x1, tss_train, features_x1_train)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m features_x1_val \u001b[39m=\u001b[39m extract_signal(bed_files_x1, tss_val, features_x1_val)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m features_x2_train \u001b[39m=\u001b[39m extract_signal(bed_files_x2, tss_train, features_x2_train)\n",
      "\u001b[1;32m/Users/huijeong/Library/Mobile Documents/com~apple~CloudDocs/ETH Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb Cell 7\u001b[0m in \u001b[0;36mextract_signal\u001b[0;34m(bed_files, tss, features)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m     signal \u001b[39m=\u001b[39m bed[\u001b[39m4\u001b[39m]\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m \u001b[39mfor\u001b[39;00m j \u001b[39min\u001b[39;00m tss\u001b[39m.\u001b[39mindex:\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39many\u001b[39;49m((\u001b[39mchr\u001b[39;49m \u001b[39m==\u001b[39;49m tss\u001b[39m.\u001b[39;49mloc[j, \u001b[39m'\u001b[39;49m\u001b[39mchr\u001b[39;49m\u001b[39m'\u001b[39;49m]) \u001b[39m&\u001b[39;49m (start \u001b[39m>\u001b[39;49m\u001b[39m=\u001b[39;49m tss\u001b[39m.\u001b[39;49mloc[j, \u001b[39m'\u001b[39;49m\u001b[39mTSS_start\u001b[39;49m\u001b[39m'\u001b[39;49m]) \u001b[39m&\u001b[39;49m (end \u001b[39m<\u001b[39;49m\u001b[39m=\u001b[39;49m tss\u001b[39m.\u001b[39;49mloc[j, \u001b[39m'\u001b[39;49m\u001b[39mTSS_end\u001b[39;49m\u001b[39m'\u001b[39;49m])):\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m         tss_signal \u001b[39m=\u001b[39m signal[(\u001b[39mchr\u001b[39m \u001b[39m==\u001b[39m tss\u001b[39m.\u001b[39mloc[j, \u001b[39m'\u001b[39m\u001b[39mchr\u001b[39m\u001b[39m'\u001b[39m]) \u001b[39m&\u001b[39m (start \u001b[39m>\u001b[39m\u001b[39m=\u001b[39m tss\u001b[39m.\u001b[39mloc[j, \u001b[39m'\u001b[39m\u001b[39mTSS_start\u001b[39m\u001b[39m'\u001b[39m]) \u001b[39m&\u001b[39m (end \u001b[39m<\u001b[39m\u001b[39m=\u001b[39m tss\u001b[39m.\u001b[39mloc[j, \u001b[39m'\u001b[39m\u001b[39mTSS_end\u001b[39m\u001b[39m'\u001b[39m])]\u001b[39m.\u001b[39msum()\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/huijeong/Library/Mobile%20Documents/com~apple~CloudDocs/ETH%20Zurich/2022-2023/ML4G/Project1/Project1_code.ipynb#X26sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m     \u001b[39melse\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# extract signal for x1 train, val, x2 train, val\n",
    "features_x1_train = pd.DataFrame()\n",
    "features_x1_val = pd.DataFrame()\n",
    "features_x2_train = pd.DataFrame()\n",
    "features_x2_val = pd.DataFrame()\n",
    "\n",
    "\n",
    "features_x1_train = extract_signal(bed_files_x1, tss_train, features_x1_train)\n",
    "features_x1_val = extract_signal(bed_files_x1, tss_val, features_x1_val)\n",
    "features_x2_train = extract_signal(bed_files_x2, tss_train, features_x2_train)\n",
    "features_x2_val = extract_signal(bed_files_x2, tss_val, features_x2_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract signal for x3\n",
    "features_x3_test = pd.DataFrame()\n",
    "features_x3_test = extract_signal(bed_files_x3, tss_x3, features_x3_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Work Package 1.2 - Model Building"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: \n",
    "# Select the best model to predict gene expression from the obtained features in WP 1.1.\n",
    "\n",
    "# ---------------------------INSERT CODE HERE---------------------------\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# ----------------------------------------------------------------------\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Work Package 1.3 - Prediction on Test Data (Evaluation Metric)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO:\n",
    "# Using the model trained in WP 1.2, make predictions on the test data (chr 1 of cell line X3).\n",
    "# Store predictions in a variable called \"pred\" which is a numpy array.\n",
    "\n",
    "pred = None\n",
    "# ---------------------------INSERT CODE HERE---------------------------\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# ----------------------------------------------------------------------\n",
    "\n",
    "# Check if \"pred\" meets the specified constrains\n",
    "assert isinstance(pred, np.ndarray), 'Prediction array must be a numpy array'\n",
    "assert np.issubdtype(pred.dtype, np.number), 'Prediction array must be numeric'\n",
    "assert pred.shape[0] == len(test_genes), 'Each gene should have a unique predicted expression'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Store Predictions in the Required Format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store predictions in a ZIP. \n",
    "# Upload this zip on the project website under \"Your submission\".\n",
    "# Zip this notebook along with the conda environment (and README, optional) and upload this under \"Your code\".\n",
    "\n",
    "save_dir = 'path/to/save/output/file'  # TODO\n",
    "file_name = 'gex_predicted.csv'         # PLEASE DO NOT CHANGE THIS\n",
    "zip_name = \"LastName_FirstName_Project1.zip\" # TODO\n",
    "save_path = f'{save_dir}/{zip_name}'\n",
    "compression_options = dict(method=\"zip\", archive_name=file_name)\n",
    "\n",
    "test_genes['gex_predicted'] = pred.tolist()\n",
    "test_genes[['gene_name', 'gex_predicted']].to_csv(save_path, compression=compression_options)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
